### PR TITLE
[Quantization] Quantized Ops

### PR DESCRIPTION
**Pull request by ZihengJiang**

### CODE REVIEW

**Comment by tqchen** [comment]
please fix the lint errors

**Comment by liangfu** [comment]
I think quantization is an interesting topic, however, it requires large mount of effort to contribute and ensure everything works. **Can we make an independent branch on dmlc/nnvm, similar like [sparse](https://github.com/dmlc/nnvm/tree/sparse), so that more people can contribute to the topic?**

**Comment by tqchen** [comment]
independent branch means divergence.  The quantization support likely does not need structure change so we will directly push for this effort in the master branch. With the help of TVM stack, the generation of quantized ops becomes easier and we already have some successful examples. 

We definitely welcome contribution and involvements. Either via issue, or discussion forum in tvm stack

File: python/nnvm/top/quantized_ops.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+@reg.register_compute("quantize")
+def compute_quantize(attrs, inputs, _):
+    """Compute definition of quantize"""
+    repr_bit = attrs.get_int('repr_bit')
+    out_dtype = attrs['out_type']
+    assert out_dtype == 'int8'
+    data = inputs[0]
+
+    bit_width = 8
+    repr_value = pow(2, repr_bit)
```
Comment (tqchen) [inline python/nnvm/top/quantized_ops.py]: use left shift

File: python/nnvm/top/tensor.py [hunk 2]
```diff
@@ -37,11 +36,42 @@ def _compute(attrs, x, _):
         return f(x[0], x[1])
     return _compute
 
-
 _fschedule_injective = tvm.convert(_schedule_injective)
 _fschedule_broadcast = _fschedule_injective
 _fschedule_elemwise = _fschedule_injective
 
+# lshift
+@reg.register_compute("left_shift")
```
Comment (tqchen) [inline python/nnvm/top/tensor.py]: this is not consistent with numpy left_shift. To make it consistent with current add, sub, we should name it ```__lshift_scalar__```

File: python/nnvm/top/tensor.py [hunk 3]
```diff
@@ -37,11 +36,42 @@ def _compute(attrs, x, _):
...
+def compute_lshift(attrs, inputs, _):
+    data = inputs[0]
+    bit = attrs.get_int("bit")
+    return topi.left_shift(data, bit)
+reg.register_pattern("left_shift", OpPattern.ELEMWISE)
+reg.register_schedule("left_shift", _fschedule_broadcast)
+
+# rshift
+@reg.register_compute("right_shift")
```
Comment (tqchen) [inline python/nnvm/top/tensor.py]: this is not consistent with numpy left_shift. To make it consistent with current add, sub, we should name it ```__rshift_scalar__```

**Comment by Laurawly** [comment]
Quantization is better to be built inside MXNet, and NNVM focuses on generating kernels. Usually in quantization process, fine-tune is required. Ideally, quantization is done in MXNet (with or without fine-tune), then use NNVM to generate lib.

**Comment by tqchen** [comment]
This is mainly for graph level and op def support for quantized models, with simple support for graph re-transformation and do not intend to serve the role of training quantized models

**Comment by liangfu** [comment]
The advantage of quantization within NNVM over MXNet, I think, is that it provides accelerated computation other than CUDA, typically for devices that support OpenCL like AMD GPUs and FPGAs. Besides, well designed neural network should be robust to quantized computation in nature. For instance, the introduction of `relu6` over `relu` would reduce the chance of numeric overflow, while still resist to degenerate accuracy.

File: include/nnvm/top/tensor.h [hunk 4]
```diff
@@ -251,6 +265,15 @@ struct ClipParam : public dmlc::Parameter<ClipParam> {
   }
 };
 
+struct AroundParam : public dmlc::Parameter<AroundParam> {
+  int mode;
+  DMLC_DECLARE_PARAMETER(AroundParam) {
+    DMLC_DECLARE_FIELD(mode);
```
Comment (etaf) [inline include/nnvm/top/tensor.h]: This is a typo? It cause compilation failure. Suggest to mv the ";" to the end of next line.

**Comment by etaf** [comment]
@ZihengJiang May I know whether this PR will be merged?

**Comment by tqchen** [comment]
c.f. #518 we will redirect further changes to tvm repo, please open a new PR there. Please invite the original set of reviewers when the new PR is opened so they can review and approve the changes
